File: src/test/java/no/nav/sosialhjelp/soknad/business/db/repositories/sendtsoknad/BatchSendtSoknadRepositoryJdbcTest.java
package no.nav.sosialhjelp.soknad.business.db.repositories.sendtsoknad;

import no.nav.sosialhjelp.soknad.business.db.RepositoryTestSupport;
import no.nav.sosialhjelp.soknad.config.DbTestConfig;
import no.nav.sosialhjelp.soknad.domain.SendtSoknad;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import javax.inject.Inject;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

import static java.time.LocalDateTime.now;
import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {DbTestConfig.class})
@ActiveProfiles("test")
class BatchSendtSoknadRepositoryJdbcTest {

    private static final String EIER = "12345678901";
    private static final String BEHANDLINGSID = "1100020";
    private static final String TILKNYTTET_BEHANDLINGSID = "4567";
    private static final String FIKSFORSENDELSEID = "12345";
    private static final String ORGNUMMER = "987654";
    private static final String NAVENHETSNAVN = "NAV Enhet";
    private static final LocalDateTime BRUKER_OPPRETTET_DATO = now().minusDays(2).truncatedTo(ChronoUnit.MILLIS);
    private static final LocalDateTime BRUKER_FERDIG_DATO = now().minusSeconds(50).truncatedTo(ChronoUnit.MILLIS);
    private static final LocalDateTime SENDT_DATO = now().truncatedTo(ChronoUnit.MILLIS);

    @Inject
    private SendtSoknadRepository sendtSoknadRepository;

    @Inject
    private BatchSendtSoknadRepository batchSendtSoknadRepository;

    @Inject
    private RepositoryTestSupport soknadRepositoryTestSupport;

    @AfterEach
    public void tearDown() {
        soknadRepositoryTestSupport.getJdbcTemplate().update("delete from SENDT_SOKNAD");
    }

    @Test
    void hentSendtSoknadHenterSendtSoknadForEierOgBehandlingsid() {
        sendtSoknadRepository.opprettSendtSoknad(lagSendtSoknad(EIER), EIER);

        Long sendtSoknadId = batchSendtSoknadRepository.hentSendtSoknad(BEHANDLINGSID).get();

        assertThat(sendtSoknadId).isNotNull();
    }

    @Test
    void slettSendtSoknadSletterSoknadFraDatabase() {
        SendtSoknad sendtSoknad = lagSendtSoknad(EIER);
        Long sendtSoknadId = sendtSoknadRepository.opprettSendtSoknad(sendtSoknad, EIER);
        sendtSoknad.setSendtSoknadId(sendtSoknadId);

        batchSendtSoknadRepository.slettSendtSoknad(sendtSoknadId);

        assertThat(batchSendtSoknadRepository.hentSendtSoknad(BEHANDLINGSID)).isEmpty();
    }

    private SendtSoknad lagSendtSoknad(String eier) {
        return new SendtSoknad().withEier(eier)
                .withBehandlingsId(BEHANDLINGSID)
                .withTilknyttetBehandlingsId(TILKNYTTET_BEHANDLINGSID)
                .withFiksforsendelseId(FIKSFORSENDELSEID)
                .withOrgnummer(ORGNUMMER)
                .withNavEnhetsnavn(NAVENHETSNAVN)
                .withBrukerOpprettetDato(BRUKER_OPPRETTET_DATO)
                .withBrukerFerdigDato(BRUKER_FERDIG_DATO)
                .withSendtDato(SENDT_DATO);
    }

}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/db/repositories/sendtsoknad/BatchSendtSoknadRepositoryJdbcTest.kt
package no.nav.sosialhjelp.soknad.db.repositories.sendtsoknad

import no.nav.sosialhjelp.soknad.Application
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import javax.inject.Inject

@ActiveProfiles(profiles = ["no-redis", "test"])
@SpringBootTest(classes = [Application::class])
internal class BatchSendtSoknadRepositoryJdbcTest {

    @Inject
    private lateinit var sendtSoknadRepository: SendtSoknadRepository

    @Inject
    private lateinit var batchSendtSoknadRepository: BatchSendtSoknadRepository

    @Inject
    private lateinit var jdbcTemplate: JdbcTemplate

    @AfterEach
    fun tearDown() {
        jdbcTemplate.update("delete from SENDT_SOKNAD")
    }

    @Test
    fun hentSendtSoknadHenterSendtSoknadForEierOgBehandlingsid() {
        sendtSoknadRepository.opprettSendtSoknad(lagSendtSoknad(EIER), EIER)
        val sendtSoknadId = batchSendtSoknadRepository.hentSendtSoknad(BEHANDLINGSID)
        assertThat(sendtSoknadId).isNotNull
    }

    @Test
    fun slettSendtSoknadSletterSoknadFraDatabase() {
        val sendtSoknad = lagSendtSoknad(EIER)
        val sendtSoknadId = sendtSoknadRepository.opprettSendtSoknad(sendtSoknad, EIER)
        sendtSoknad.sendtSoknadId = sendtSoknadId!!
        batchSendtSoknadRepository.slettSendtSoknad(sendtSoknadId)
        assertThat(batchSendtSoknadRepository.hentSendtSoknad(BEHANDLINGSID)).isNull()
    }

    private fun lagSendtSoknad(eier: String): SendtSoknad {
        return SendtSoknad(
            behandlingsId = BEHANDLINGSID,
            tilknyttetBehandlingsId = TILKNYTTET_BEHANDLINGSID,
            eier = eier,
            fiksforsendelseId = FIKSFORSENDELSEID,
            orgnummer = ORGNUMMER,
            navEnhetsnavn = NAVENHETSNAVN,
            brukerOpprettetDato = BRUKER_OPPRETTET_DATO,
            brukerFerdigDato = BRUKER_FERDIG_DATO,
            sendtDato = SENDT_DATO
        )
    }

    companion object {
        private const val EIER = "12345678901"
        private const val BEHANDLINGSID = "1100020"
        private const val TILKNYTTET_BEHANDLINGSID = "4567"
        private const val FIKSFORSENDELSEID = "12345"
        private const val ORGNUMMER = "987654"
        private const val NAVENHETSNAVN = "NAV Enhet"
        private val BRUKER_OPPRETTET_DATO = LocalDateTime.now().minusDays(2).truncatedTo(ChronoUnit.MILLIS)
        private val BRUKER_FERDIG_DATO = LocalDateTime.now().minusSeconds(50).truncatedTo(ChronoUnit.MILLIS)
        private val SENDT_DATO = LocalDateTime.now().truncatedTo(ChronoUnit.MILLIS)
    }
}
